java-checks/src/main/java/org/sonar/java/checks/spring/SpringScanDefaultPackageCheck.java

```diff
@@ -0,0 +1,171 @@
+
+  private static final Map<String, Set<String>> SCAN_PACKAGE_ATTRIBUTES = buildScanPackageAttributes();
+
+  private static Map<String, Set<String>> buildScanPackageAttributes() {
+    Map<String, Set<String>> map = new HashMap<>();
+
+    map.put("org.springframework.context.annotation.ComponentScan",
+      new HashSet<>(Arrays.asList(DEFAULT_ATTRIBUTE, "basePackages", "basePackageClasses")));
+
+    map.put("org.springframework.boot.autoconfigure.SpringBootApplication",
+      new HashSet<>(Arrays.asList("scanBasePackages", "scanBasePackageClasses")));
+
+    map.put("org.springframework.boot.web.servlet.ServletComponentScan",
+      new HashSet<>(Arrays.asList(DEFAULT_ATTRIBUTE, "basePackages", "basePackageClasses")));
+
+    return map;
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.ANNOTATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    AnnotationTree annotation = (AnnotationTree) tree;
+    Set<String> scanPackageAttributeNames = SCAN_PACKAGE_ATTRIBUTES.get(annotation.symbolType().fullyQualifiedName());
+    if (scanPackageAttributeNames != null) {
+      List<ExpressionTree> scanPackageAttributes = annotation.arguments().stream()
+        .filter(argument -> scanPackageAttributeNames.contains(attributeName(argument)))
+        .collect(Collectors.toList());
+
+      if (scanPackageAttributes.isEmpty()) {
+        checkIfAnnotationIsInDefaultPackage(annotation);
+      } else {
+        checkAnnotationPackageAttributes(annotation, scanPackageAttributes);
+      }
+    }
+  }
+
+  private void checkIfAnnotationIsInDefaultPackage(AnnotationTree annotation) {
+    if (isNodeInDefaultPackage(annotation)) {
+      reportIssue(annotation.annotationType(), "Remove the annotation \"@" + annotation.symbolType().name() +
+        "\" or move the annotated class out of the default package.");
+    }
+  }
+
+  private void checkAnnotationPackageAttributes(AnnotationTree annotation, List<ExpressionTree> scanPackageAttributes) {
+    List<ExpressionTree> values = scanPackageAttributes.stream()
+      .flatMap(SpringScanDefaultPackageCheck::extractValues)
+      .collect(Collectors.toList());
+
+    values.stream()
+      .flatMap(SpringScanDefaultPackageCheck::findEmptyString)
+      .forEach(expression -> reportIssue(expression, "Define packages to scan. Don't rely on the default package."));
+
+    values.stream()
+      .flatMap(SpringScanDefaultPackageCheck::findClassInDefaultPackage)
+      .forEach(identifier -> reportIssue(identifier, "Remove the annotation \"@" + annotation.symbolType().name() +
+        "\" or move the \"" + identifier.name() + "\" class out of the default package."));
+  }
+
+  private static String attributeName(ExpressionTree expression) {
+    if (expression.is(Tree.Kind.ASSIGNMENT)) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) expression;
+      if (assignment.variable().is(Tree.Kind.IDENTIFIER)) {
```

Comment:
can it be something else ? is it covered ? if not and it is defensive programming : worth a comment to explain lack of coverage.

Reply:
the only possible kind for "variable()" in annotation is IDENTIFIER ([JavaGrammar.java:L589](https://github.com/SonarSource/sonar-java/blob/01ec449f34ed81b46a85c30a1df89131d2432ac2/java-frontend/src/main/java/org/sonar/java/ast/parser/JavaGrammar.java#L589)) , I removed the test.